code/math/vecmat.cpp

```diff
@@ -2571,6 +2571,56 @@ void vm_vec_boxscale(vec2d *vec, float  /*scale*/)
 	vec->y *= ratio;
 }
 
+// adds one matrix to another.
+// dest can equal source
+// dest += src
+void vm_matrix_add2(matrix* dest, const matrix* src)
+{
+	dest->vec.fvec += src->vec.fvec;
+	dest->vec.rvec += src->vec.rvec;
+	dest->vec.uvec += src->vec.uvec;
+}
+
+// TODO Remove this function if we ever move to a math library like glm
+/**
+* @brief							Attempts to invert a 3x3 matrix
+* @param[inout]		dest     		The inverted matrix, or nullptr if inversion is impossible
+* @param[in]		m   			Pointer to the matrix we want to invert
+*
+* @returns							Whether or not the matrix is invertible
+*/
+bool vm_inverse_matrix(matrix* dest, const matrix* m)
+{
+	matrix inv;	// create a temp matrix so we can avoid getting a determinant that is 0
+
+	// Use a2d so it's easier for people to read
+	inv.a2d[0][0] = -m->a2d[1][2] * m->a2d[2][1] + m->a2d[1][1] * m->a2d[2][2];
+	inv.a2d[0][1] = m->a2d[0][2] * m->a2d[2][1] - m->a2d[0][1] * m->a2d[2][2];
+	inv.a2d[0][2] = -m->a2d[0][2] * m->a2d[1][1] + m->a2d[0][1] * m->a2d[1][2];
+	inv.a2d[1][0] = m->a2d[1][2] * m->a2d[2][0] - m->a2d[1][0] * m->a2d[2][2];
+	inv.a2d[1][1] = -m->a2d[0][2] * m->a2d[2][0] + m->a2d[0][0] * m->a2d[2][2];
+	inv.a2d[1][2] = m->a2d[0][2] * m->a2d[1][0] - m->a2d[0][0] * m->a2d[1][2];
+	inv.a2d[2][0] = -m->a2d[1][1] * m->a2d[2][0] + m->a2d[1][0] * m->a2d[2][1];
+	inv.a2d[2][1] = m->a2d[0][1] * m->a2d[2][0] - m->a2d[0][0] * m->a2d[2][1];
+	inv.a2d[2][2] = -m->a2d[0][1] * m->a2d[1][0] + m->a2d[0][0] * m->a2d[1][1];
+
+	float det = m->a2d[0][0] * inv.a2d[0][0] + m->a2d[0][1] * inv.a2d[1][0] + m->a2d[0][2] * inv.a2d[2][0];
+	if (det == 0) {
+		dest = nullptr;
```

Comment:
Setting `dest = nullptr` is not going to have any non-local effects.  Instead, the destination matrix should be set to some trivial matrix analogous to the trivial vector in `vm_vec_normalize`.  (I notice that the matrix4 version of this function has the same bug.)